hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/ClusterJoinDelayTest.java

```diff
@@ -0,0 +1,83 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.internal.cluster.impl;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.TcpIpConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.spi.properties.ClusterProperty;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.OverridePropertyRule;
+import static com.hazelcast.spi.properties.ClusterProperty.WAIT_SECONDS_BEFORE_JOIN_ASYNC;
+import static com.hazelcast.test.OverridePropertyRule.set;
+import static com.hazelcast.test.TestEnvironment.HAZELCAST_TEST_USE_NETWORK;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+/**
+ * Make sure there are no delays when joining the cluster
+ *
+ * @author lprimak
+ */
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(QuickTest.class)
+public class ClusterJoinDelayTest extends HazelcastTestSupport {
+    @Rule
+    public final OverridePropertyRule overridePropertyRule = set(HAZELCAST_TEST_USE_NETWORK, "true");
+    @Rule
+    public final OverridePropertyRule overridePropertyRule2 = set(WAIT_SECONDS_BEFORE_JOIN_ASYNC.getName(), "true");
+
+    private TestHazelcastInstanceFactory fact;
+    private final int numInstances = 3;
+
+    @Before
+    public void beforeRun() {
+        fact = createHazelcastInstanceFactory(numInstances);
+    }
+
+    @After
+    public void afterRun() {
+        fact.shutdownAll();
+    }
+
+    @Override
+    protected Config getConfig() {
+        Config config = new Config();
+        // make sure the wait is longer than the tested-for delay.
+        // here we make sure that the newHazelcastInstance() call returns w/o blocking
+        config.setProperty(ClusterProperty.WAIT_SECONDS_BEFORE_JOIN.getName(), "120");
+        config.setProperty(ClusterProperty.MAX_WAIT_SECONDS_BEFORE_JOIN.getName(), "480");
+        TcpIpConfig tcpIpConfig = new TcpIpConfig();
+        tcpIpConfig.setEnabled(true).addMember("localhost");
+        config.getNetworkConfig().setPublicAddress("localhost").getJoin().setTcpIpConfig(tcpIpConfig);
+        return config;
+    }
+
+    @Test(timeout = 30 * 1000)
+    public void noBlockingBeforeJoin() {
+        HazelcastInstance hz1 = fact.newHazelcastInstance(getConfig());
+        HazelcastInstance hz2 = fact.newHazelcastInstance(getConfig());
+        HazelcastInstance hz3 = fact.newHazelcastInstance(getConfig());
+        assertClusterSizeEventually(numInstances, hz1, hz2, hz3);
```

Comment:
An alternative to needing a timeout on the test itself would be specifying the `timeout` parameter in `assertClusterSizeEventually()`

Reply:
Didn't know that took timeout... lemme look

Reply:
done